### ISSUE TITLE
fmpz allocation is broken in current trunk

### ISSUE THREAD

**Issue by tthsqe12** [wbhart/flint2#1088]
@albinahlback
aors_ui.c
We cannot realloc the _mp_d directly and have to go through gmp's mpz layer to realloc it (sorry).
something like
mf->_mp_d = flimbs = FLINT_MPZ_REALLOC(mf, gabssz)
Though, I will leave it to you to decide how to change it.
If you want to see trunk failing, just try starting the nemo test code :)

**Comment by albinahlback** [comment]
Perhaps it would be a good idea to include checking Nemo in our workflows.

**Comment by albinahlback** [comment]
I mean, `FLINT_MPZ_REALLOC` basically evaluates to `realloc`, and I was told by Bill to use `flint_realloc` instead.

**Comment by albinahlback** [comment]
It was some time ago that I was writing interfaces between C and Julia. How do I use other binaries than what is given by `FLINT_JLL`? If I recall correctly, something was written under Nemo's documentation, but I cannot find it.

**Comment by tthsqe12** [comment]
https://github.com/Nemocas/Nemo.jl/blob/master/README.maintainer.md

**Comment by tthsqe12** [comment]
And FLINT_MPZ_REALLOC does _not_ just evaluate to {flint_}realloc

**Comment by albinahlback** [comment]
> https://github.com/Nemocas/Nemo.jl/blob/master/README.maintainer.md

Thanks!

> And FLINT_MPZ_REALLOC does _not_ just evaluate to {flint_}realloc

As I see it:

1. `FLINT_MPZ_REALLOC` calls `_mpz_realloc`,
2. `_mpz_realloc` calls `__GMP_REALLOCATE_FUNC_LIMBS`,
3. `__GMP_REALLOCATE_FUNC_LIMBS` is a macro that calls `__GMP_REALLOCATE_FUNC_TYPE`,
4. `__GMP_REALLOCATE_FUNC_TYPE` is a macro that evaluates to `__gmp_reallocate_func`,
5. `__gmp_reallocate_func` is defined in `memory.c` as `__gmp_default_reallocate`,
6. `__gmp_default_reallocate` is a function that calls `realloc` (and asserts that the return value is not `NULL`).

Perhaps there is some step here that I have missed, or perhaps other programs redefines `__gmp_reallocate_func`.

**Comment by tthsqe12** [comment]
step 2. is: calls `__GMP_ALLOCATE_FUNC_LIMBS` or `__GMP_REALLOCATE_FUNC_LIMBS`

**Comment by albinahlback** [comment]
Okay, sounds like I have to add `flint_malloc` to the mix.

**Comment by albinahlback** [comment]
In PR #1077, we can see in the [CI for Nemo.jl at line 173](https://github.com/wbhart/flint2/runs/5806063644?check_suite_focus=true#step:8:173) that it indeed fails for `_fmpz_add_mpn_1`. It seems to call `flint_realloc`, which calls `jl_realloc`, which seems to be part of the garbage collector.

It seems like this order should be correct. Will keep investigating.

**Comment by albinahlback** [comment]
If you see [this run](https://github.com/albinahlback/flint2/runs/5821131924?check_suite_focus=true), where I have changed `flint_realloc` to simply `realloc`, it does not fail at that step anymore; it can do alot more tests without failing.

**Comment by albinahlback** [comment]
There the problems seems to be `nf_init`.

**Comment by albinahlback** [comment]
The problem is
```julia
R, x = PolynomialRing(QQ, :x)
K, a = NumberField(x^3 + 3 * x + 1, :a)
```

**Comment by thofma** [comment]
Did you rebuild antic? There are some binary incompatible changes in flint 2.9, which makes it crash with our prebuilt antic binaries.

**Comment by albinahlback** [comment]
Ah... Okay, will try this. Thanks!

**Comment by albinahlback** [comment]
Okay, yeah, it works. Thanks Tommy!

**Comment by albinahlback** [comment]
I think the best solution is to use GMP's allocation methods. Looking at GMP_jll's [`build_tarballs.jl`-ish](https://github.com/JuliaPackaging/Yggdrasil/blob/1fe18336ec2b0a9dc634785b89f1004eb81205e4/G/GMP/common.jl), we see that, in contrast to [FLINT_jll's](https://github.com/JuliaPackaging/Yggdrasil/blob/750e6c51e56a196c1343befd00e1bbf0a60fd8d9/F/FLINT/build_tarballs.jl), they do not override their allocation functions. And since `_fmpz_mpn_add_1` really use wants to mimic GMP, I think the best way is to use `__gmp_reallocate_func` defined in `gmp-impl.h`.

**Comment by thofma** [comment]
They do it here: https://github.com/JuliaLang/julia/blob/master/base/gmp.jl#L103-L107

**Comment by albinahlback** [comment]
Oh, thanks again! But anyway, they redefine `__gmp_reallocate_func`. And we want to use that function. (I meant `__gmp_default_reallocate` earlier, which wouldn't work in case as `__gmp_reallocate_func` wouldn't point there.)

**Comment by fredrik-johansson** [comment]
We use ``mpz_realloc2`` elsewhere when we need to reallocate an ``mpz``. Why not here?

**Comment by albinahlback** [comment]
It is slower as it's one more function call. `_fmpz_add_mpn_1` is meant to be a low-level function.

**Comment by albinahlback** [comment]
I can obtain the functions via `mp_get_memory_functions` defined in `gmp.h` (i.e. a visible function, intended for outside users), which could be inside `memory_manager.c` in FLINT.

**Comment by fredrik-johansson** [comment]
But when a reallocation is actually needed, the reallocation itself is going to be much more expensive than the function call overhead.

**Comment by wbhart** [comment]
Unless I am mistaken, this is satisfactorily resolved in #1093 

Feel free to reopen the issue if you believe this is incorrect.
